WordPress/Classes/Services/JetpackSocialService.swift

```diff
@@ -0,0 +1,77 @@
+import WordPressKit
+import CoreData
+
+class JetpackSocialService {
+
+    // MARK: Properties
+
+    private let coreDataStack: CoreDataStackSwift
+
+    private lazy var remote: JetpackSocialServiceRemote = {
+        let api = coreDataStack.performQuery { context in
+            return WordPressComRestApi.defaultV2Api(in: context)
+        }
+        return .init(wordPressComRestApi: api)
+    }()
+
+    // MARK: Methods
+
+    init(coreDataStack: CoreDataStackSwift = ContextManager.shared) {
+        self.coreDataStack = coreDataStack
+    }
+
+    /// Fetches and updates the Publicize information for the site associated with the `blogID`.
+    /// The method returns a value type that contains the remaining usage of Social auto-sharing and the maximum limit for the associated site.
+    ///
+    /// - Note: If the returned result is a success with nil sharing limit, it's likely that the blog is hosted on WP.com, and has no Social sharing limitations.
+    ///
+    /// Furthermore, even if the sharing limit exists, it may not be applicable for the blog since the user might have purchased a product that ignores this limitation.
+    ///
+    /// - Parameters:
+    ///   - blogID: The ID of the blog.
+    ///   - completion: Closure that's called after the sync process completes.
+    func syncSharingLimit(for blogID: Int, completion: @escaping (Result<PublicizeInfo.SharingLimit?, Error>) -> Void) {
+        remote.fetchPublicizeInfo(for: blogID) { [weak self] result in
+            switch result {
+            case .success(let remotePublicizeInfo):
+                self?.coreDataStack.performAndSave({ context -> PublicizeInfo.SharingLimit? in
+                    guard let blog = try Blog.lookup(withID: blogID, in: context) else {
+                        // unexpected to fall into this case, since the API should return an error response.
+                        throw ServiceError.blogNotFound(id: blogID)
+                    }
```

Comment:
Would there be a risk of a crash if this were hit? The chances are low but I'm thinking it'd be possible to land in this scenario with:

- Slow internet connection and/or the service taking a long time
- While the network call is still in flight: 
  - User deletes their blog
  - User logs out of the app

Additionally, what are your thoughts on updating the `try` here to be `try?` in case that lookup throws?

Reply:
> Would there be a risk of a crash if this were hit?

No, any errors thrown from within this specific `performAndSave` block will be caught and converted into a `failure`, and get passed into the `completion` parameter of the method:

https://github.com/wordpress-mobile/WordPress-iOS/blob/1d267828267666e768531567eed085ec16c0b0f2/WordPress/Classes/Utility/ContextManager.swift#L129-L141

(Notice line 133: The `block` parameter is executed within `Result(catching:)` which allows us to catch and convert any thrown errors into a failure result.)

> what are your thoughts on updating the `try` here to be `try?` in case that lookup throws?

It's a bit tricky because this is a three-valued logic situation:
1. API request succeeded, with `PublicizeInfo` fields included. This is likely what's returned for Jetpack sites.
2. API request succeeded, without `PublicizeInfo` fields. This is the case for WP.com sites.
3. An error occurred. 

This is the reason why the result type is `<SharingLimit?, Error>`. If we return `nil` from the guard condition's `else` block, it'll blur between situations no. 2 and 3.

Reply:
> No, any errors thrown from within this specific performAndSave block will be caught and converted into a failure, and get passed into the completion parameter of the method

I missed this part, thanks for explanation!

> This is the reason why the result type is <SharingLimit?, Error>. If we return nil from the guard condition's else block, it'll blur between situations no. 2 and 3.

Makes sense. I'm good with leaving it as is then 👍